### ISSUE TITLE
Implement the memory module with backward()

### ISSUE THREAD

**Issue by RPegoud** [RPegoud/Hercules#3]
Following the procedure outlined in the Titans paper, we compute the memory updates using:
```python
grad_fn = torch.func.grad(self._associative_memory_loss, has_aux=True)
        per_chunk_grad_fn = vmap(grad_fn, in_dims=(None, 1, 1, 1))
        per_chunk_grads, associative_loss = per_chunk_grad_fn(
            dict(params), k, v, adaptive_lr
        )
        per_chunk_grads = TensorDict(per_chunk_grads)

        self.associative_loss = associative_loss

        with torch.no_grad():  # Disable grad for test-time updates
            for name, param in self.lmm.named_parameters():
                if per_chunk_grads.get(name) is not None:
                    grad = per_chunk_grads[name].mean(dim=0)  # average per chunk
                    momentum = momentum_dict[name].clone()

                    alpha_t = adaptive_forget.mean(dim=[1, 2, 3])
                    theta_t = adaptive_lr.mean(dim=[1, 2, 3])
                    eta_t = adaptive_momentum.mean(dim=[1, 2, 3])

                    alpha_t = flatten_and_expand(alpha_t, grad.dim())
                    eta_t = flatten_and_expand(eta_t, grad.dim())
                    theta_t = flatten_and_expand(theta_t, grad.dim())

                    # Momentum update: S_t = η_t * S_{t-1} - θ_t * grad
                    momentum.mul_(eta_t)
                    momentum.sub_(theta_t * grad)

                    # Parameter update: M_t = (1 - α_t) * M_{t-1} + S_t
                    new_param = param * (1.0 - alpha_t) + momentum
                    param.data.copy_(new_param.mean(0))

                    buffer_name = f"momentum_{name.replace('.', '_')}"
                    getattr(self, buffer_name).copy_(momentum.mean(dim=0))
```
This computes and applies the gradients manually and relies on a function call to the model:
```python
def _associative_memory_loss(
        self,
        params: TensorDict,
        inputs: torch.Tensor,
        targets: torch.Tensor,
        adaptive_lr: torch.Tensor,
    ) -> tuple[torch.Tensor, torch.Tensor]:
        preds = functional_call(self.lmm, params, inputs)
        loss = F.mse_loss(preds, targets, reduction="none").mean(dim=-1)
        weighted_loss = loss * adaptive_lr.squeeze()
        total_loss = weighted_loss.sum()
        return total_loss, loss
```
This functional call and the fact that llama is frozen imply that pytorch won't track any gradients. This prevents `accelerate` from being able to synchronise training on multiple GPUs. 

Looking at the ablation study in the original paper, I decide to revert back to a standard training loop on `backward()`, using `adamW` as optimiser to enable parallelised training.

<img width="474" alt="Image" src="https://github.com/user-attachments/assets/ce851bbc-ac71-4983-8dd0-1e4a900b4f9e" />
